File: catalyst_voices/packages/internal/catalyst_voices_repositories/lib/src/api/models/payment_data.dart
import 'package:json_annotation/json_annotation.dart';

part 'payment_data.g.dart';

/// A role payment address information.
@JsonSerializable()
final class PaymentData {
  /// Indicates if the data is persistent or volatile.
  @JsonKey(name: 'is_persistent')
  final bool isPersistent;

  /// A time when the address was added.
  final DateTime time;

  /// Cardano Blockchain Slot Number
  /// A block slot number.
  final int slot;

  /// Transaction Index
  @JsonKey(name: 'txn_index')
  final int txnIndex;

  /// Cardano Payment Address
  /// An option payment address.
  /// CIP-19 - Cardano Addresses - https://cips.cardano.org/cip/CIP-19
  final String? address;

  const PaymentData({
    required this.isPersistent,
    required this.time,
    required this.slot,
    required this.txnIndex,
    this.address,
  });

  factory PaymentData.fromJson(Map<String, dynamic> json) => _$PaymentDataFromJson(json);

  Map<String, dynamic> toJson() => _$PaymentDataToJson(this);
}

File: catalyst-gateway/bin/src/service/api/cardano/rbac/registrations_get/v2/payment_data.rs
//! A role payment address information.

use cardano_chain_follower::pallas_addresses::ShelleyAddress;
use poem_openapi::{Object, types::Example};

use crate::service::common::types::{
    cardano::{cip19_shelley_address::Cip19ShelleyAddress, slot_no::SlotNo, txn_index::TxnIndex},
    generic::{boolean::BooleanFlag, date_time::DateTime},
};

/// A role payment address information.
#[derive(Object, Debug, Clone)]
#[oai(example)]
pub struct PaymentData {
    /// Indicates if the data is persistent or volatile.
    is_persistent: BooleanFlag,
    /// A time when the address was added.
    time: DateTime,
    /// A block slot number.
    slot: SlotNo,
    /// A transaction index.
    txn_index: TxnIndex,
    /// An option payment address.
    address: Option<Cip19ShelleyAddress>,
}

impl PaymentData {
    /// Creates a new `PaymentData` instance.
    pub fn new(
        is_persistent: bool,
        time: DateTime,
        slot: SlotNo,
        txn_index: TxnIndex,
        address: Option<ShelleyAddress>,
    ) -> anyhow::Result<Self> {
        let address = address.map(Cip19ShelleyAddress::try_from).transpose()?;

        Ok(Self {
            is_persistent: is_persistent.into(),
            time,
            slot,
            txn_index,
            address,
        })
    }
}

impl Example for PaymentData {
    fn example() -> Self {
        Self {
            is_persistent: BooleanFlag::example(),
            time: DateTime::example(),
            slot: SlotNo::example(),
            txn_index: TxnIndex::example(),
            address: Some(Cip19ShelleyAddress::example()),
        }
    }
}
